Update code to support Mojo and Max 25.1

msaelices opened the pull request:
- [X] Adapt constructors to the out keyword format
- [X] In new Mojo the arguments are not implicitly converted. We need a @implicit constructor
- [X] Adapt the deprecated "-> T as out:" to use the out keyword
- [X] Adapt removed _lit_mut_cast which now is Origin.cast_from
- [X] Adapt Type{field1: 42, field2: 17} syntax removed from Mojo 25.1 into equivalent one
- [X] inout -> mut replacements
- [X] Some fixes to adhere to Mojo 25.1
- [X] Adapt mojoproject.toml
- [X] Prepare buffers OP. 
- [X] Echo server example

msaelices committed:
Some fixes to adhere to Mojo 25.1

msaelices committed:
inout -> mut replacements

msaelices committed:
Adapt Type{field1: 42, field2: 17} syntax removed from Mojo 25.1 into equivalent one

msaelices committed:
Adapt removed _lit_mut_cast which now is Origin.cast_from

msaelices committed:
Adapt the deprecated "-> T as out:" to use the out keyword

msaelices committed:
In new Mojo the arguments are not implicitly converted. We need a @implicit constructor

msaelices committed:
Adapt remaining Type{field1: 42, field2: 17} syntax removed from Mojo 25.1 into equivalent on

msaelices committed:
Adapt constructors to the out keyword format

msaelices committed:
Fix the mojix code to support Max & Mojo 25.1

msaelices committed:
build.sh script to build the packages

msaelices committed:
OwnedFd is a trivial type to be copied and moved, added @value

msaelices committed:
Script for running the tests. Remove __init__.mojo to fix the test discovery

msaelices committed:
Add queue example

msaelices committed:
Add WIP echoserver example

msaelices committed:
Implement PrepProvideBuffers op, equivalent to io_uring_prep_provide_buffers

msaelices committed:
Use different buffers per client connected

msaelices committed:
Clean the echoserver example fixing some edge cases

msaelices committed:
More edge cases fixed

msaelices committed:
Some more fixes related to buffers in the echoserver example

msaelices committed:
Use the bid declaration instead of conn.bid in echoserver

dmitry-salin commented on mojix/fd.mojo:
```diff
@@ -170,6 +170,7 @@ struct IoUringFd[is_registered: Bool](FileDescriptor, IoUringFileDescriptor):
         return self
 
 
+@value
```
`OwnedFd` must not have a copy constructor, or the type becomes an unsafe abstraction rather than a safe one. Imagine a situation where, after calling `__del__` on one instance, the operating system reuses its id for another open resource, and then calling `__del__` on the second instance closes it.

msaelices replied:
Good catch! Removed here: https://github.com/msaelices/io_uring/commit/3a5bf27c80bbe7a3ba2316a00f53031c3c5c0fcf

dmitry-salin commented on mojix/io_uring.mojo:
```diff
@@ -413,38 +425,39 @@ struct IoUringFeatureFlags(Defaultable, Boolable):
         return self.value != 0
 
 
+@value
```
`IoUringRegisterOp` and similar types are used to emulate enums. Thus, there should be as few ways as possible to construct directly from an arbitrary `UInt32` value. In this case infallible construction from integer is unsafe. I think it is better to use:
```
@always_inline("nodebug")
fn __init__(out self, *, unsafe_id: UInt32):
    self.id = unsafe_id
```

msaelices replied:
I agree. Fixed: https://github.com/msaelices/io_uring/commit/6712a6380b321b82bd91d831f22c1881164e4b2e

Thanks!

dmitry-salin replied:
I think that adding a overloaded constructor does not prevent the `@value` decorator from synthesizing the default one. The point is to not have a default constructor, because it is unsafe in this case. This also applies to all other "enum" types that used curly braces initializer, except `PollingMode`, `SQE` and `CQE`.

msaelices replied:
I'm not sure if I understood it correctly but we cannot construct structs without passing the explicit `unsafe_id=` keyword param. You can see it here in one line I changed to expose the compiler error:

![image](https://github.com/user-attachments/assets/fb0b5c72-5bc4-4610-9747-96b03c4165ac)

dmitry-salin replied:
Ok, I just wasn't sure if this overload prevents the @value decorator from synthesizing the default constructor.

dmitry-salin commented on mojoproject.toml:
```diff
@@ -1,6 +1,6 @@
 [project]
 authors = ["Dmitry Salin <dmitry.salin@gmail.com>"]
-channels = ["conda-forge", "https://conda.modular.com/max-nightly"]
+channels = ["conda-forge", "https://conda.modular.com/max"]
```
Mojo is currently far from "feature complete" enough to tie the `main` branch to release versions. I think it's better to use a separate `stable` branch for this.

msaelices replied:
I kind of agree, but I also wanted to add the library to the Modular channel to be called from other libraries like my https://github.com/msaelices/mojo-websockets/ because my problem is that I don't want to wait for the proper async support in Mojo to be able to have concurrent connections.

And the issue with Modular official channel is that AFAIK it only supports the stable Max versions (see https://docs.google.com/document/d/1l0BE3lvNJPCs4vXXHpVFzv4Z1ebj3ODknMyqQ5pyxvM/edit?tab=t.0)

Then, if you merge this PR in your repo, it will be more updated as the 25.1 is newer than the current `24.6.0.dev2024111816` one.

Then, we can create the stable branch from the main one and the main one would be pointing to the `25.2.0.dev` matching the nightly version from Modular.

What do you think?

dmitry-salin replied:
Ok, let's remove the requirement for nightly from the `README.md` in this PR then.

dmitry-salin replied:
Also this repository will eventually be split into 3 libraries: `linux_raw`, `mojix`, `io_uring`.

dmitry-salin commented on run_tests.mojo:
```diff
@@ -1,8 +0,0 @@
-fn main() raises:
-    test_nop()
-    test_nop_skip_cqe()
-    test_accept_timeout()
```
This test failed when run through the test framework. The problem was a thread blocking when calling `enter` in the kernel. The test library crashed for some unknown reason. It looks like this issue has been fixed.

msaelices replied:
Yes, I have no problem now:

![image](https://github.com/user-attachments/assets/d6b9317e-58ad-49a4-91a6-62847e7648e7)

dmitry-salin commented on io_uring/op.mojo:
```diff
@@ -521,3 +521,68 @@ struct Write[type: SQE, origin: MutableOrigin](Operation):
+
+
+@register_passable
+struct PrepProvideBuffers[type: SQE, origin: MutableOrigin](Operation):
```
In general I am not against adding any operations, but this can be considered a legacy op. The functionality of registered buffers has changed over time (leaving a mess behind): https://github.com/tokio-rs/tokio-uring/issues/112#issuecomment-1247280807

A newer type, called ring mapped buffers is supported since 5.19.
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=c7fb19428d67dd0a2a78a4f237af01d39c78dc5a

https://github.com/axboe/liburing/wiki/io_uring-and-networking-in-2023#provided-buffers

This functionality deserves a carefully designed, safe abstraction. Something like https://github.com/tokio-rs/io-uring/pull/256 is planned.

msaelices replied:
Oh! I did not know that! Thanks for the heads-up. 

I just needed the OP because my echoserver.mojo example was based in this version in C: https://github.com/frevib/io_uring-echo-server/blob/master/io_uring_echo_server.c

So:

1. If possible, I would like to keep this old OP considering as a legacy one just to be able to move ahead with my work in mojo-websockets, except if the new ring mapped buffers would be implemented soon.
2. I would love to help you implementing the new ring mapped buffers. As you can guess, I'm not an expert in io_uring but, I would love to learn about it.

msaelices replied:
@dmitry-salin I have this WIP PR with the ring mapped buffers support: https://github.com/msaelices/io_uring/pull/1

The PR is targeting my own `main` branch until we finally merge this PR into your repo.

dmitry-salin replied:
@msaelices I think this PR can be split and examples added later. This is a new library and I think it's good to have examples that use modern approaches. We can add ring mapped buffers and examples later. I'll check your branch and compare it to what I had in mind.

dmitry-salin replied:
I implemented ring mapped buffers in my Rust library earlier. I may try to port it to Mojo this weekend.

msaelices replied:
@dmitry-salin removed the examples from this PR.

msaelices committed:
Fix the edge case of only echoing back the first message

msaelices committed:
Removed the @value from OwnedFd to prevent race conditions when copying the FD

msaelices committed:
Explicit constructor to IoUringRegisterOp to prevent implicit creations from regular UInt32 values

msaelices commented:
@dmitry-salin thanks for your review. Could you please take another look?

dmitry-salin commented on README.md:
```diff
@@ -15,18 +15,16 @@ For Mojo LSP to work correctly in editors like [`Helix`](https://github.com/heli
 
-
 ## Requirements
 Currently only x86_64 platform is supported. Mojo nightly builds are required.
```
```suggestion
Currently only x86_64 platform is supported.
```

msaelices committed:
Remove the examples

msaelices committed:
Fix issue in the README as we are supporting the stable Mojo version

Co-authored-by: Dmitry Salin <dmitry.salin@gmail.com>

msaelices commented:
@dmitry-salin could you please take a look again?

Also, I've implemented an echoserver example in another PR, which makes no use of the legacy registered buffers op, so, if you want, I can get rid of it in this PR.

You can take a look here: https://github.com/msaelices/io_uring/pull/2

dmitry-salin commented:
> @dmitry-salin could you please take a look again?
> 
> Also, I've implemented an echoserver example in another PR, which makes no use of the legacy registered buffers op, so, if you want, I can get rid of it in this PR.
> 
> You can take a look here: [msaelices#2](https://github.com/msaelices/io_uring/pull/2)

I think we can add legacy operations later if they are absolutely needed for something. And then `ProvideBuffers` should be added with a corresponding `RemoveBuffers` operation.

dmitry-salin commented on io_uring/utils.mojo:
```diff
@@ -2,12 +2,13 @@ from sys.intrinsics import _RegisterPackType, llvm_intrinsic
-    alias ACQUIRE = Self {id: 0}
-    alias RELEASE = Self {id: 1}
-    alias RELAXED = Self {id: 2}
+    alias ACQUIRE = Self(id=0)
```
```suggestion
    alias ACQUIRE = Self(unsafe_id=0)
```
All types that emulate enums should not have a default constructor, as this is unsafe.
Better add:
```
@always_inline("nodebug")
fn __init__(out self, *, unsafe_id: UInt8):
    self.id = unsafe_id
```

msaelices replied:
Fixed: https://github.com/msaelices/io_uring/commit/e3f8457df8005e0f83323e1e00c31b826e62a268

dmitry-salin commented on mojix/io_uring.mojo:
```diff
@@ -589,57 +621,58 @@ struct IoUringCqeFlags(Defaultable, Boolable):
-    alias URING_CMD = Self {id: IORING_OP_URING_CMD}
-    alias SEND_ZC = Self {id: IORING_OP_SEND_ZC}
-    alias SENDMSG_ZC = Self {id: IORING_OP_SENDMSG_ZC}
+    alias NOP = Self(id=IORING_OP_NOP)
```
```suggestion
    alias NOP = Self(unsafe_id=IORING_OP_NOP)
```
All types that emulate enums should not have a default constructor, as this is unsafe.

msaelices replied:
Fixed: https://github.com/msaelices/io_uring/commit/e3f8457df8005e0f83323e1e00c31b826e62a268

dmitry-salin commented on mojix/io_uring.mojo:
```diff
@@ -656,10 +689,11 @@ struct IoUringFsyncFlags(Defaultable):
 struct IoUringMsgRingCmds:
-    alias DATA = Self {id: IORING_MSG_DATA}
-    alias SEND_FD = Self {id: IORING_MSG_SEND_FD}
+    alias DATA = Self(id=IORING_MSG_DATA)
```
```suggestion
    alias DATA = Self(unsafe_id=IORING_MSG_DATA)
```
All types that emulate enums should not have a default constructor, as this is unsafe.

msaelices replied:
Fixed: https://github.com/msaelices/io_uring/commit/e3f8457df8005e0f83323e1e00c31b826e62a268

dmitry-salin commented on mojix/mm.mojo:
```diff
@@ -209,34 +219,35 @@ struct ProtFlags(Defaultable):
-    alias POPULATE_WRITE = Self {id: MADV_POPULATE_WRITE}
-    alias DONTNEED_LOCKED = Self {id: MADV_DONTNEED_LOCKED}
-    alias COLLAPSE = Self {id: MADV_COLLAPSE}
+    alias NORMAL = Self(id=MADV_NORMAL)
```
```suggestion
    alias NORMAL = Self(unsafe_id=MADV_NORMAL)
```
All types that emulate enums should not have a default constructor, as this is unsafe.

msaelices replied:
Fixed: https://github.com/msaelices/io_uring/commit/e3f8457df8005e0f83323e1e00c31b826e62a268

dmitry-salin commented on mojix/net/types.mojo:
```diff
@@ -194,11 +194,11 @@ alias RawSocketType = c_uint
-    alias SEQPACKET = Self {id: SOCK_SEQPACKET}
-    alias RAW = Self {id: SOCK_RAW}
-    alias RDM = Self {id: SOCK_RDM}
+    alias STREAM = Self(id=SOCK_STREAM)
```
```suggestion
    alias STREAM = Self(unsafe_id=SOCK_STREAM)
```
All types that emulate enums should not have a default constructor, as this is unsafe.

msaelices replied:
Fixed: https://github.com/msaelices/io_uring/commit/e3f8457df8005e0f83323e1e00c31b826e62a268

dmitry-salin commented on mojix/net/types.mojo:
```diff
@@ -238,11 +243,11 @@ alias RawAddrFamily = __kernel_sa_family_t
-    alias INET6 = Self {id: AF_INET6}
-    alias NETLINK = Self {id: AF_NETLINK}
-    alias UNIX = Self {id: AF_UNIX}
+    alias UNSPEC = Self(id=AF_UNSPEC)
```
```suggestion
    alias UNSPEC = Self(unsafe_id=AF_UNSPEC)
```
All types that emulate enums should not have a default constructor, as this is unsafe.

msaelices replied:
Fixed: https://github.com/msaelices/io_uring/commit/e3f8457df8005e0f83323e1e00c31b826e62a268

dmitry-salin commented on mojix/net/types.mojo:
```diff
@@ -252,37 +257,37 @@ struct AddrFamily:
-    alias ICMPV6 = Self {id: IPPROTO_ICMPV6}
-    alias MH = Self {id: IPPROTO_MH}
-    alias ROUTING = Self {id: IPPROTO_ROUTING}
+    alias IP = Self(id=IPPROTO_IP)
```
```suggestion
    alias IP = Self(unsafe_id=IPPROTO_IP)
```
All types that emulate enums should not have a default constructor, as this is unsafe.

msaelices replied:
Fixed: https://github.com/msaelices/io_uring/commit/e3f8457df8005e0f83323e1e00c31b826e62a268

dmitry-salin replied:
@msaelices this type remains unchanged in the last commit.

msaelices replied:
The problem with this type is that it's `Defaultable`. So I need to remove the trait to make this useful.

msaelices replied:
Done: https://github.com/msaelices/io_uring/commit/e3f8457df8005e0f83323e1e00c31b826e62a268

msaelices committed:
Add linux_raw package to the build.sh script

msaelices committed:
Remove the PROVIDE_BUFFERS op

msaelices commented:
> > @dmitry-salin could you please take a look again?
> > Also, I've implemented an echoserver example in another PR, which makes no use of the legacy registered buffers op, so, if you want, I can get rid of it in this PR.
> > You can take a look here: [msaelices#2](https://github.com/msaelices/io_uring/pull/2)
> 
> I think we can add legacy operations later if they are absolutely needed for something. And then `ProvideBuffers` should be added with a corresponding `RemoveBuffers` operation.

Removed here: https://github.com/msaelices/io_uring/commit/6bc983dc76b1e97c5cfee4941fef1514899530e0

msaelices committed:
Remove default constructors from all enum emulating structs

msaelices commented:
@dmitry-salin I think all the requested changes are done. Could you please re-review?

msaelices committed:
Remove Defaultable trait from Protocol, with unsafe_id constructor
